src/Effect/Ref.purs

```diff
@@ -2,6 +2,38 @@
 -- |
 -- | _Note_: `Control.Monad.ST` provides a _safe_ alternative to `Ref` when
 -- | mutation is restricted to a local scope.
+-- |
+-- | You'll notice that all of the functions that operate on a `Ref`
+-- | (e.g. `new`, `modify`, `modify_`, `read`) return an `Effect (Ref s)`
+-- | and not a `Ref s`. Here's what would happe if the `Ref s` was not
+-- | wrapped in an `Effect`:
+-- | 1. In some situations, one could violate referential transparency.
+-- | 2. One can violate the type system by giving a `Ref` a polymorphic type
```

Comment:
After reflecting on this a bit more, I think this particular point _is_ specific to `Ref.new`, and is probably best documented there (sorry).